DM-43586: Add versioning to the fits module

arunkannawadi opened the pull request.

arunkannawadi commented on python/lsst/cell_coadds/_fits.py:
```diff
@@ -40,13 +51,25 @@
 from lsst.geom import Box2I, Extent2I, Point2I
 from lsst.obs.base.formatters.fitsGeneric import FitsGenericFormatter
 from lsst.skymap import Index2D
+from packaging import version
```
Strictly speaking, this is not explicitly listed as. dependency in `rubin-env`, but exists in our environment. I've seen it used in `Spectractor` and `ctrl_bps` as well. Alternative would be to move `VersionTuple` from `daf_butler` to a more generic utils and reuse it from there.

TallJimbo replied:
I think that means it should get an RFC (and should have when it was first used by `ctrl_bps`; I'm not sure about the rules for `Spectractor`).  Moving the butler stuff to `utils` seems fine as well, if @timj agrees.

timj replied:
We also use `packaging` in `lsst-versions`. It seems like the thing we are recommended to use these days to parse version. Square also use it. It seems to be used pervasively given a quick scan of my site-packages (although pip, setuptools, and wheel all vendor it internally). I will make a note on the v9 rubin-env ticket.

TallJimbo commented on python/lsst/cell_coadds/_fits.py:
```diff
@@ -19,11 +19,22 @@
+readable or not. Any changes to this module, even as trivial as formatting
+changes and documentation updates should bump the version number (in this case,
+it would bump the patch). This is mandated in GitHub Actions.
+"""
```
I don't think it's obvious how one maps semver to I/O versioning; I think we need to spell out what is and is not supported under major, minor, and patch version bumps (and we might find that those are not sufficiently granular for the guarantees we might want to make, if both forwards- and backwards-compatibility and read/write permutations are in play).

But I also think going to all that effort might be premature, and maybe a single-integer version would be better for now?  The most important thing is getting some kind of number into the files so future code doesn't have to guess about what it's dealing with.

I also have to admit that I'm a little skeptical of the value of the GitHub Action and the policy of bumping the patch version for all changes.  Is this something you've seen in other codebases?

arunkannawadi replied:
I'm envisioning the semantic versioning to work in exactly the same way as it does for software. I suppose one assumption here is that the version with which we are reading a file is never older than the version used to write a file, since a write operation must preceed a read operation. An example of a major bump would be if we decide to use `ImageHDU`s instead of `BinTableHDU`s and drop the support for the latter. That would be a breaking change rendering the old files unreadable. Examples of minor bump would include this PR, supporting versions with or without `VERSION` in the header and substituting a default value if not found for newer items like `day_obs` and `physical_filter` we just introduced. The changes in all of the three examples aren't fundamentally different - in one case _we decide_ to drop support, and for the others we don't.

arunkannawadi replied:
We still rely on the developer doing the due diligence in deciding the correct version after a change. This is also enforceable but much harder. We could also call `0.x.y` versions unstable in that we don't give any semver kind of guarantee until `1.0.0`. Bumping minor versions for every change would be equivalent to just an integer numbering that you suggested.  It seems wasteful to me to discard old data for any missing metadata. So this is definitely something we should support in production, even if not now.

arunkannawadi commented:
This is ready for another round of review, Jim.

TallJimbo commented on python/lsst/cell_coadds/_fits.py:
```diff
@@ -75,11 +110,61 @@ def __init__(self, filename: str) -> None:
+            return False
+
+        if this_version.minor < other_version.minor:
+            return False
```
This feels too pessimistic.  It's shouldn't be that hard for us to ignore header keys, table columns, and HDUs we don't recognize.

TallJimbo commented on python/lsst/cell_coadds/_fits.py:
```diff
@@ -75,11 +110,61 @@ def __init__(self, filename: str) -> None:
+                "Minor version of the %s does not match the version on the file, for major version 0."
+                "The read is not guaranteed to be successful",
+                cls.__name__,
+            )
```
I don't think we want the warning here; we can and should warn about specific things we're dropping on the floor, but warning when there's a good chance everything is fine is a bit too alarmist.

TallJimbo reviewed (commented):
I'm afraid it seems we didn't land on the same page as much as we both thought after talking about it live.  I'd still like to at least hope for a bit more forward-compatibility, and I'm still unconvinced about "semantic versioning" and especially "patch versions" being useful concepts here.  (Major and minor versions are great, but I think semantic versioning implies a lot more than that.)

arunkannawadi commented:
> I'm afraid it seems we didn't land on the same page as much as we both thought after talking about it live.

My bad. My scant notes of our discussion partly to blame here. Let's revisit this again next week. I think versioning is important since this is a dataset that is likely to be bulk-downloaded and accessed outside of RSP so I'd like to get a good versioning scheme going on.

arunkannawadi commented:
I've switched to using just major and minor versions, and removed any mention of semantic versioning. I've expanded on the policy on bumping versions and compatibility across versions is largely put in by hand. I'm hoping to wrap this up, barring any minor implementation changes after today.

TallJimbo commented on doc/changes/README.rst:
```diff
@@ -0,0 +1,24 @@
+
+An example file name would therefore look like ``DM-30291.misc.rst``.
+
+If the change concerns specifically the registry or a datastore the news fragment can be placed in the relevant subdirectory.
```
```suggestion
```

TallJimbo commented on python/lsst/cell_coadds/_fits.py:
```diff
@@ -19,14 +19,57 @@
+module constant FILE_FORMAT_VERSION. The policy for bumping the version is:
+
+1. When the on-disk format written by this module changes such that the
+accompanying reader can still read files written by the previous version, then
```
I think minor bumps are when older readers can still read files written by new writers.  As you've described it here, the _format_ is not changing; only the reader code is.  If you don't want to try to rigorously support compatibility in that direction (as suggested in the examples), then we could say "old readers _may_ be able to read new code", while major would mean they definitely cannot.

TallJimbo commented on python/lsst/cell_coadds/_fits.py:
```diff
@@ -69,17 +125,71 @@ class CellCoaddFitsReader:
+            cls.MAXIMUM_FILE_FORMAT_VERSIONS,
+            strict=True,
+        ):
+            if min_version <= written_version < max_version:
```
I didn't realize this kind of construct actually worked in Python.  Are you sure it does?

arunkannawadi replied:
Yes, it does! Here's some random blog from 12 years ago: https://www.johndcook.com/blog/2012/04/03/a-b-c/

And by `a <= b < c`, I do mean `a < b` and `b < c` which is exactly how Python interprets it as well.

TallJimbo replied:
:exploding_head:

arunkannawadi replied:
Oh, I should also note that I do mean to compare strings here, not after casting them to `float`.

TallJimbo commented on python/lsst/cell_coadds/_fits.py:
```diff
@@ -19,14 +19,57 @@
+
+It is plausible that different (non-consequent) major versions are compatible
+with each other (due to reverting back to an earlier format, or to something
+very similar). Such cases should be marked in the `isCompatibleWith` method.
```
I would describe `isCompatibleWith` as testing whether a particular format version can be read by the current reader (such that the answer may be "yes" due to intentional backwards-compatibility).  It is not a measure of compatibility between formats.

TallJimbo reviewed (commented):
I'm still not fully on board with the descriptions of what major and minor version bumps mean, though I do agree with your examples.  So I'm hoping we're almost on the same page now.

TallJimbo reviewed (approved):
Discussed this live, and we're pretty much on the same page; we'll leave major vs. minor to developer discretion, with _intent_ to _possibly_ deprecate being a strong indication that a major version bump is in order.

arunkannawadi commented:
Thinking about this more after the call, I am with your points stronger than I was during the call. I'll clean up the changes and merge by EOD tomorrow. Thank you for all the discussions, Jim.
